### PR TITLE
Add mapMaybe functions for Map and Set types, filter for Set types

### PR DESCRIPTION
**Pull request by garyb**

### CODE REVIEW

File: src/Data/Map/Internal.purs [hunk 1]
```diff
@@ -644,3 +646,13 @@ filterKeys predicate = filterWithKey $ const <<< predicate
...
+
+-- | Applies a function to each key/value pair in a map, discarding entries
+-- | where the function returns `Nothing`.
+mapMaybeWithKey :: forall k a b. Ord k => (k -> a -> Maybe b) -> Map k a -> Map k b
+mapMaybeWithKey f = foldrWithIndex (\k a acc → maybe acc (\b -> insert k b acc) (f k a)) empty
+
+-- | Applies a function to each value in a map, discarding entries where the
+-- | function returns `Nothing`.
+mapMaybe :: forall k a b. Ord k => (a -> Maybe b) -> Map k a -> Map k b
+mapMaybe f = foldrWithIndex (\k a acc → maybe acc (\b -> insert k b acc) (f a)) empty
```
Comment (kritzcreek) [inline src/Data/Map/Internal.purs]: What about `mapMaybe f = mapMaybeWithKey (\_ -> f)` here?

File: src/Data/Set/NonEmpty.purs [hunk 2]
```diff
@@ -147,3 +149,13 @@ properSubset (NonEmptySet s1) (NonEmptySet s2) = Set.properSubset s1 s2
...
+
+-- | Filter out those values of a set for which a predicate on the value fails
+-- | to hold.
+filter :: forall a. Ord a => (a -> Boolean) -> NonEmptySet a -> Set a
+filter f (NonEmptySet s) = Set.filter f s
+
+-- | Applies a function to each value in a set, discarding entries where the
+-- | function returns `Nothing`.
+mapMaybe :: forall a b. Ord b => (a -> Maybe b) -> NonEmptySet a -> Set b
+mapMaybe f (NonEmptySet s) = Set.mapMaybe f s
```
Comment (garyb) [inline src/Data/Set/NonEmpty.purs:161]: Maybe these should return `Maybe NonEmptySet`?

**Review by kritzcreek — APPROVED** [review]
Great, let's make sure to open an issue asking for an optimized version after we merge this

File: src/Data/Set/NonEmpty.purs (hunk 2 above)
Comment (garyb) [reply src/Data/Set/NonEmpty.purs:161]: I'm not sure though, some of the functions in here work that way, but in other NonEmptyXs we just lower to possibly-empty for situations like this.

Comment (kritzcreek) [reply src/Data/Set/NonEmpty.purs:161]: I'd say going to `Set` is the right way to go. It's how we do `filter` on NonEmptyList etc.

Comment (garyb) [reply src/Data/Set/NonEmpty.purs:161]: Makes sense 👍

**Pull request merged**